### PR TITLE
Improve perf of Add/Attach/Update by caching key metadata

### PR DESCRIPTION
**Pull request by ajcvickers**
Properties keep track of which keys they belong to.

### CODE REVIEW

**Comment by ajcvickers** [comment]
Note: Diego and Rowan probably need to review this so people can hold off reviewing for now.

File: src/EntityFramework.Core/Metadata/Internal/IPropertyKeyMetadata.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+using System.Collections.Generic;
+using JetBrains.Annotations;
+
+namespace Microsoft.Data.Entity.Metadata.Internal
+{
+    public interface IPropertyKeyMetadata
+    {
+        IKey PrimaryKey { get; [param: CanBeNull] set; }
+        IReadOnlyList<IKey> Keys { get; [param: CanBeNull] set; }
+        IReadOnlyList<IForeignKey> ForeignKeys { get; [param: CanBeNull] set; }
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/IPropertyKeyMetadata.cs]: Why are the setters part of the interface?

Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/Internal/IPropertyKeyMetadata.cs]: Left over from previous slightly different implementation. I will remove them.

**Comment by AndriySvyryd** [comment]
:shipit:

**Pull request merged**